x/wasm/ibc_test.go

```diff
@@ -4,11 +4,125 @@ import (
+	"github.com/cosmos/cosmos-sdk/types/address"
 	clienttypes "github.com/cosmos/ibc-go/v4/modules/core/02-client/types"
 	channeltypes "github.com/cosmos/ibc-go/v4/modules/core/04-channel/types"
+	ibcexported "github.com/cosmos/ibc-go/v4/modules/core/exported"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/rand"
+
+	"github.com/CosmWasm/wasmd/x/wasm/keeper"
+	"github.com/CosmWasm/wasmd/x/wasm/types"
 )
 
+func TestOnRecvPacket(t *testing.T) {
+	anyRelayerAddr := sdk.AccAddress(rand.Bytes(address.Len))
+	anyContractIBCPkg := IBCPacketFixture(func(p *channeltypes.Packet) {
+		p.DestinationPort = "wasm.cosmos1w09vr7rpe2agu0kg2zlpkdckce865l3zps8mxjurxthfh3m7035qe5hh7f"
+	})
+	myCustomEvent := sdk.NewEvent("testing")
+	specs := map[string]struct {
+		ibcPkg      channeltypes.Packet
+		contractRsp ibcexported.Acknowledgement
+		contractErr error
+		expEvents   sdk.Events
+		expPanic    bool
+		expAck      ibcexported.Acknowledgement
+	}{
+		"contract returns success response": {
+			ibcPkg:      anyContractIBCPkg,
+			contractRsp: keeper.ContractConfirmStateAck([]byte{1}),
+			expAck:      keeper.ContractConfirmStateAck([]byte{1}),
+			expEvents: sdk.Events{
+				myCustomEvent,
+				{
+					Type: "ibc_packet_received",
+					Attributes: []abci.EventAttribute{
+						{Key: []byte("module"), Value: []byte("wasm")},
+						{Key: []byte("_contract_address"), Value: []byte("cosmos1w09vr7rpe2agu0kg2zlpkdckce865l3zps8mxjurxthfh3m7035qe5hh7f")},
+						{Key: []byte("success"), Value: []byte("true")},
+					},
+				},
+			},
+		},
+		"contract returns err response": {
+			ibcPkg:      anyContractIBCPkg,
+			contractRsp: channeltypes.NewErrorAcknowledgement(types.ErrInvalid.Wrap("testing")),
+			expAck:      channeltypes.NewErrorAcknowledgement(types.ErrInvalid.Wrap("testing")),
+			expEvents: sdk.Events{
+				{
+					Type: "ibc_packet_received",
+					Attributes: []abci.EventAttribute{
+						{Key: []byte("module"), Value: []byte("wasm")},
+						{Key: []byte("_contract_address"), Value: []byte("cosmos1w09vr7rpe2agu0kg2zlpkdckce865l3zps8mxjurxthfh3m7035qe5hh7f")},
+						{Key: []byte("success"), Value: []byte("false")},
+					},
+				},
+			},
+		},
+		"nil considered success response": { // regression only
+			ibcPkg: anyContractIBCPkg,
+			expEvents: sdk.Events{
+				myCustomEvent,
+				{
+					Type: "ibc_packet_received",
+					Attributes: []abci.EventAttribute{
+						{Key: []byte("module"), Value: []byte("wasm")},
+						{Key: []byte("_contract_address"), Value: []byte("cosmos1w09vr7rpe2agu0kg2zlpkdckce865l3zps8mxjurxthfh3m7035qe5hh7f")},
+						{Key: []byte("success"), Value: []byte("true")},
+					},
+				},
+			},
+		},
+		"unknown contract port": {
+			ibcPkg: IBCPacketFixture(func(p *channeltypes.Packet) {
+				p.DestinationPort = "not-a-contract-port"
+			}),
+			expPanic: true,
+		},
+		"contract executed with error": {
```

Comment:
I'm confused here.
I thought this was exactly the case we wanted to panic?

Oh wait, this mocks out the keeper and the panic happens inside the keeper. Nevermind, I think this is good. (Just not sure when we return such error)

Reply:
This gives a wrong impression, indeed.
It is not the contract that returns the error (as capture this and panic) but a message execution that was returned with `Res.Ok.Messages` by the contract.

Reply:
Okay, so a better title is what is needed here?